public/gameInfo.php

```diff
@@ -300,32 +300,133 @@ function drawCharts() {
 
         dataTotalScore.addRows([
             <?php
-            $largestWonByCount = 0;
-            $count = 0;
-            $plural = '';
-            for ($i = 1; $i <= $numAchievements; $i++) {
-                if ($count++ > 0) {
-                    $plural = 's';
-                    echo ", ";
+            function calculateBuckets(int $numAchievements, array $achDist, array $achDistHardcore): array {
+                // Enable bucketing based on the number of achievements in the set.
+                // This number was picked arbitrarily, but generally reflects when we start seeing
+                // width constraints in the Achievements Distribution bar chart.
+                $isBucketingEnabled = $numAchievements > 60;
+
+                // Bucket size is determined based on the total number of achievements in the set.
+                // If bucketing is enabled, we aim for roughly 20 buckets (hence dividing by 20).
+                // If bucketing is not enabled, each achievement gets its own bucket (bucket size is 1).
+                $bucketSize = $isBucketingEnabled ? ceil($numAchievements / 20) : 1;
```

Comment:
Using integer sized buckets leads to some awkward distributions. Here's what the buckets look like for a game with 78 achievements (Snake's Revenge-NES):
![image](https://github.com/RetroAchievements/RAWeb/assets/32680403/e33e2ade-dfed-4cf0-bbb8-d7baf0b4f9f4)

The last few buckets are:
65-68=4
69-72=4
73-76=4
77-77=1
78-78=1

78/4 = 3.9, so the buckets should be 0, 3.9, 7.8, 11.7, 15.6, 19.5, 23.4, 27.3, 31.2, 35.1, 39, 42.9, 46.8, 50.7, 54.6, 58.5, 62.4, 66.3, 70.2, 74.1, 78. You would round those values to get a more fluid distribution of buckets.

Additionally, are you're explicitly keeping a separate bucket for "all achievements", the calculation should be `($numAchievements - 1) / 19`, which would yield the following 20 buckets: 4, 8, 12, 16, 20, 24, 28, 32, 36, 41, 45, 49, 53, 57, 61, 65, 69, 73, 77, 78. The first 19 buckets are evenly distributed using rounded values (note the small jump between buckets 36 and 41 as the fractional portion transitions), and the 20th bucket is reserved for the "all achievements".

Reply:
Good catch, this was difficult indeed to work through, but I may have landed on a resolution in https://github.com/RetroAchievements/RAWeb/pull/1607/commits/1f430f4964997718468b477db45824746e9fb8fb.

First of all, thank you for providing the Snake's Revenge example. This was an excellent test case.

I've adjusted how "smoothing" between the buckets works for hopefully better distributions to be presented to the user:

<img width="184" alt="Screenshot 2023-06-11 at 1 53 39 PM" src="https://github.com/RetroAchievements/RAWeb/assets/3984985/f7e6a0bf-9f9a-4e9e-9c2d-1c9050c429ec">

<img width="174" alt="Screenshot 2023-06-11 at 1 53 45 PM" src="https://github.com/RetroAchievements/RAWeb/assets/3984985/aca96b1e-8fee-4ccf-801c-61e0bafd9d4c">


I _think_ I've got a better handle on this now, but always possible I've missed something. Let me know if you spot anything that looks bizarre.

Reply:
The X axis labels go from 1-20 now. Can we get it to show start values instead (since the last bucket will always be N, it would be better to have the second-to-last bucket not be N-1, and then the first bucket would be 1, providing the even spacing all the way to the final bucket.

Reply:
Oops, I deleted something I shouldn't have. As of https://github.com/RetroAchievements/RAWeb/pull/1607/commits/aff4a526ecf6ffbf47d7e973e60009bcf5672607 this functionality should be restored.

Reply:
I still think it would be better to use the bucket start as the X axis labels (particularly for the bookend entries [4/77 in this case would become 1/74]). It would be a smoother distribution with the last bucket (78).
![image](https://github.com/RetroAchievements/RAWeb/assets/32680403/a61101ec-cd55-48ce-b3f0-d664b51a4b61)

Also, it might be beneficial to use an odd number of buckets so the last one gets labelled.

Reply:
Both of these pieces of feedback should now be applied.

For the odd number of buckets, I increased the bucket count by 1 rather than decreased. When decreasing by 1, every single bar got a label on the x-axis which felt kind of cramped.